### ISSUE TITLE
Some queries are returning errors

### ISSUE THREAD

**Issue by ViVa98** [maoosi/prisma-appsync#16]
I'm using this prisma-appsync to deploy GraphQL API to lambda and AppSync using my existing PostgreSQL database.  Some queries are returning errors while executing. I'm using the command `introspect` initially to get `prisma.schema` based on the database.

```
ERROR	Invoke Error 	{
    "errorType": "TypeError",
    "errorMessage": "Cannot read property 'findUnique' of undefined",
    "stack": [
        "TypeError: Cannot read property 'findUnique' of undefined",
        "    at ut.get (/var/task/prisma/generated/prisma-appsync/client/index.js:2:72228)",
        "    at async $s.resolve (/var/task/prisma/generated/prisma-appsync/client/index.js:2:76727)",
        "    at async Runtime.exports.handler (/var/task/index.js:18:18)"
    ]
}
```

**Comment by maoosi** [comment]
Hey @ViVa98. From the error message, it seems like it is failing when trying to execute a get operation against a particular model ([see source code here](https://github.com/maoosi/prisma-appsync/blob/9ee3d115b1e5f8cec9fe6a8874855ae36ca04990/src/client/_resolver.ts#L404)).

Prisma-AppSync Client act as a wrapper around Prisma Client, so assuming you were trying to query `getUser`, the underlying method is trying to execute `await prisma.user.findUnique()` from inside your Lambda function. And the error would mean that `user` is undefined inside your Prisma Client.

From there, my best guess would be that somehow your generated Prisma Client (generated via `npx prisma generate` and deployed with your Lambda function) would be out-of-sync with the GraphQL Schema and Resolvers Config currently deployed on your AppSync instance.

To confirm this is indeed the issue, you can try to insert `app.prisma.user` (replacing user with your failing model name) just after `const app = new PrismaAppSync(...)` inside your Lambda function handler. When using VSCode, you should see a Type error appear:

> Property 'user' does not exist on type 'PrismaClient<PrismaClientOptions, never, RejectOnNotFound | RejectPerOperation>'.ts(XXX)

If you are using the provided CDK Boilerplate for deployment, it should automatically run `npx prisma generate` during bundling - as long as you have the right setup inside your `schema.prisma` file (make sure it wasn't removed after running the `introspect` command):

```prisma
generator client {
    provider = "prisma-client-js"
    binaryTargets = ["native", "rhel-openssl-1.0.x"]
}

generator appsync {
    provider = "prisma-appsync"
}
```

Let me know if that help you figure out the issue?

If not, please share your Prisma Schema (removing sensitive data, if any) and more details around your setup and deploy process (e.g. Are you using AWS CDK and the [provided boilerplate](https://prisma-appsync.vercel.app/guides/installation.html#using-prisma-appsync-boilerplate) to deploy your API?).

**Comment by ViVa98** [comment]
I got your concern about model is undefined in prisma client but in my local system I'm able to get data using prisma client in vs code terminal.

**Comment by ViVa98** [comment]
One more thing is, in file cdk/index.ts line no 122 `forceDockerBundling: true` is causing issues if there is no docker installed in local system. If the local system OS is windows, some commands like `cp` `rm` in the same above mentioned file also causing issues for command `cdk deploy`

**Comment by maoosi** [comment]
When using Docker for bundling, the commands inside `commandHooks` (cdk/index.ts line 101) are to copy your schema at the right place inside the running Docker instance, execute `npx prisma generate` (to make sure your schema and AppSync files are in sync) and bundle your Lambda function with the correct runtime libraries and Prisma Client.

If you had to change some of these to work locally on Windows, then it is possible that Prisma Client is not compiled/bundled properly because of it. I am working on mac, so it will be hard for me to help with this. Are you able to install and run Docker on your machine instead?

**Comment by DregondRahl** [comment]
There are two ways to handle this on windows.
1. if you have git bash, use that for the cli to install and deploy
2. Use WSL ubuntu2 for the install and deploy (ubuntu will need a change to the prisma client file names in the cdk file)

Docker bundling is not required for CDK. You can run it without and get the same outcome, if it throws any errors, as I remember inside the AWS CDK code its looking if "esbuild"  is installed to skip docker build.

> 
> 
> One more thing is, in file cdk/index.ts line no 122 `forceDockerBundling: true` is causing issues if there is no docker installed in local system. If the local system OS is windows, some commands like `cp` `rm` in the same above mentioned file also causing issues for command `cdk deploy`

**Comment by ViVa98** [comment]
I tried using docker as well. Still getting the same error. And also I can see my models suggesting by vscode in handler file for `app.prisma.people` 

> (property) PrismaClient<Prisma.PrismaClientOptions, never, Prisma.RejectOnNotFound | Prisma.RejectPerOperation>.people: Prisma.peopleDelegate<Prisma.RejectOnNotFound | Prisma.RejectPerOperation>

> prisma.people: Exposes CRUD operations for the people model. Example usage:

> // Fetch zero or more People
> const people = await prisma.people.findMany()

Here is the schema I'm using 
```
model people {
  id            Int             @id @default(autoincrement())
  name          String?         @db.VarChar
  country       String?         @db.VarChar
  person_emails person_emails[]
}

model person_emails {
  id        Int     @id @default(autoincrement())
  person_id Int
  value     String? @db.VarChar
  people    people  @relation(fields: [person_id], references: [id])

  @@index([person_id], name: "idx_fk_person_id")
}
```
I'm using `AWS RDS PostgreSQL` as a database. 

@DregondRahl @maoosi if either of you can use the above schema, deploy it to AWS and check why it is happening.

**Comment by ViVa98** [comment]
@DregondRahl @maoosi @ryands17 
I found something strange in logs of appsync and lambda in CloudWatch. Here is what I found
```
INFO	Received event :

"info": {
        "fieldName": "listperson_emails",
        "selectionSetList": [
            "value"
        ],
        "selectionSetGraphQL": "{\n  value\n}",
        "parentTypeName": "Query",
        "variables": {}
    },
```

If you see in the `fieldName` there is `listperson_email` where `person_email` is the model name.

```
INFO	Adapter output :

{
    "operation": "list",
    "model": "personEmail",
    "args": {
        "select": {
            "value": true
        },
        "skip": 0,
        "take": 50
    }
}
```
But in the adapter output, it is taking model as `personEmail` instead of `person_email`.

```
INFO	Running beforeResolve hook on "list/personEmail" with params :

{
    "authIdentity": {
        "authorization": "API_KEY"
    },
    "operation": "list",
    "subject": "PersonEmail",
    "fields": [
        "value"
    ],
    "requestSetPaths": [
        "list/personEmail/value"
    ],
    "args": {
        "select": {
            "value": true
        },
        "skip": 0,
        "take": 50
    }
}
```

Because of this model name mismatch, the `Cannot read property 'findMany' of undefined` error is returning.

**Comment by ryands17** [comment]
Could you share a basic reproduction of your repo so that I can deploy it myself and check?

**Comment by ViVa98** [comment]
https://github.com/ViVa98/PrismaAppSync

**Comment by maoosi** [comment]
I know why this is happening. There is a function in Prisma-AppSync that converts model names to PascalCase in order to convert a model `Post` to `listPosts` for example. And this has only been tested for model names without the "_" symbol.

As a quick fix, I would suggest that you rename your models as following:
1. Rename `people` to `People`
2. Rename `person_emails` to `PersonEmails`

This should do the trick (let us know if not) until we can find a longer-term fix.

**Comment by ViVa98** [comment]
Yeah, I was thinking of renaming model names using `@map` and `@@map` as mentioned on this page
> https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/use-custom-model-and-field-names#using-map-and-map-to-rename-fields-and-models-in-the-prisma-client-api

**Comment by ViVa98** [comment]
> I know why this is happening. There is a function in Prisma-AppSync that converts model names to PascalCase in order to convert a model `Post` to `listPosts` for example. And this has only been tested for model names without the "_" symbol.
> 
> As a quick fix, I would suggest that you rename your models as following:
> 
> 1. Rename `people` to `People`
> 2. Rename `person_emails` to `PersonEmails`
> 
> This should do the trick (let us know if not) until we can find a longer-term fix.

After renaming model names to camelCase, API is working now. But somehow still if model name is `People`,
```
INFO	Received event :

"info": {
        "selectionSetList": [
            "name",
            "country"
        ],
        "selectionSetGraphQL": "{\n  name\n  country\n}",
        "parentTypeName": "Mutation",
        "fieldName": "createPeople",
        "variables": {}
    },
```
```
INFO	Adapter output :

{
    "operation": "create",
    "model": "person",
    "args": {
        "select": {
            "name": true,
            "country": true
        },
        "data": {
            "name": "John Doe",
            "country": "IN"
        }
    }
}
```
```
INFO	Running beforeResolve hook on "create/person" with params :

{
    "authIdentity": {
        "authorization": "API_KEY"
    },
    "operation": "create",
    "subject": "Person",
    "fields": [
        "name",
        "country"
    ],
    "requestSetPaths": [
        "create/person/name",
        "create/person/country"
    ],
    "args": {
        "select": {
            "name": true,
            "country": true
        },
        "data": {
            "name": "John Doe",
            "country": "IN"
        }
    }
}
```
@maoosi 
Just a heads-up for you to check this whenever you're working on this bug.

**Comment by ViVa98** [comment]
> > I know why this is happening. There is a function in Prisma-AppSync that converts model names to PascalCase in order to convert a model `Post` to `listPosts` for example. And this has only been tested for model names without the "_" symbol.
> > As a quick fix, I would suggest that you rename your models as following:
> > 
> > 1. Rename `people` to `People`
> > 2. Rename `person_emails` to `PersonEmails`
> > 
> > This should do the trick (let us know if not) until we can find a longer-term fix.
> 
> After renaming model names to camelCase, API is working now. But somehow still if model name is `People`,
> 
> ```
> INFO	Received event :
> 
> "info": {
>         "selectionSetList": [
>             "name",
>             "country"
>         ],
>         "selectionSetGraphQL": "{\n  name\n  country\n}",
>         "parentTypeName": "Mutation",
>         "fieldName": "createPeople",
>         "variables": {}
>     },
> ```
> 
> ```
> INFO	Adapter output :
> 
> {
>     "operation": "create",
>     "model": "person",
>     "args": {
>         "select": {
>             "name": true,
>             "country": true
>         },
>         "data": {
>             "name": "John Doe",
>             "country": "IN"
>         }
>     }
> }
> ```
> 
> ```
> INFO	Running beforeResolve hook on "create/person" with params :
> 
> {
>     "authIdentity": {
>         "authorization": "API_KEY"
>     },
>     "operation": "create",
>     "subject": "Person",
>     "fields": [
>         "name",
>         "country"
>     ],
>     "requestSetPaths": [
>         "create/person/name",
>         "create/person/country"
>     ],
>     "args": {
>         "select": {
>             "name": true,
>             "country": true
>         },
>         "data": {
>             "name": "John Doe",
>             "country": "IN"
>         }
>     }
> }
> ```
> 
> @maoosi
> Just a heads-up for you to check this whenever you're working on this bug.

@maoosi
Above mentioned problem is related to the usage of singular and plural forms of a particular word.
- If the model name is defined with the name `People`, at the time of creating the .gql file singular form of `People` which is `Person` is using for single type operations. 
- For other operations which are used to do bulk things like deleting all records, getting all records, `People` is taking into consideration.

**Comment by maoosi** [comment]
Thanks @ViVa98 that helps a lot! I've started working on a fix, that consist of generating a new mapping config, for Prisma-AppSync to be able to keep track of original Prisma Client model names.

It will be included with the next Beta release.

**Comment by maoosi** [comment]
Resolved in [1.0.0-beta.58.2](https://github.com/maoosi/prisma-appsync/releases/tag/1.0.0-beta.58.2) (Fix: Issue linked to Prisma models naming").
